### ISSUE TITLE
Regression: False positive E1111: Assigning to function call which doesn't return (assignment-from-no-return)

### ISSUE THREAD

**Issue by sscherfke** [PyCQA/pylint#2385]
### Steps to reproduce
Assign return value of a “special” function to a name, e.g. the `q = ...` in the following example:
```python
def decorate(fn):
    """Decorate *fn* to return ``self`` to enable chained method calls."""
    def wrapper(self, *args, **kw):
        fn(self, *args, **kw)
        return self
    return wrapper


class Test:
    @decorate
    def meth(self):
        pass


def main():
    Test().meth().meth()
    t = Test()
    t = t.meth()  # Bang!
    assert isinstance(t, Test)


if __name__ == '__main__':
    main()
```
A behavior like this is used, e.g., by sqlalchemys `Query` class (`Query().filter(...).filter(...).all()`)

### Current behavior
```
************* Module e1111
/tmp/e1111.py:12:4: E1111: Assigning to function call which doesn't return (assignment-from-no-return)

-------------------------------------------------------------------
Your code has been rated at 5.45/10
```

### Expected behavior
```

---------------------------------------------------------------------
Your code has been rated at 10.00/10
```

### pylint --version output

```
pylint 2.1.0
astroid 2.0.2
Python 3.6.5 | (default, Aug  3 2018, 08:59:28)
[GCC 4.8.5 20150623 (Red Hat 4.8.5-28)]
```
It works with 
```
pylint 1.9.3,
astroid 1.6.5
```

**Comment by pbasista** [comment]
It would help to narrow down the example, so that it is easy to see exactly which “special” function property is causing Pylint to incorrectly interpret assignment as an error. Preferably, such an example should not use any 3rd party library.

**Comment by sscherfke** [comment]
I updated the example to not use any third party packages.

**Comment by pbasista** [comment]
Interesting. Pylint 2 probably considers function or method as returning `None` immediately when the function's body returns `None`. It seems like the decorators are not taken into account.

**Comment by sscherfke** [comment]
Yes.  I just tested it with pylint/astroid == 2.0.0 and the issue occurs there, too. So it was introduced between 1.9.3 and 2.0.0.

**Comment by pbasista** [comment]
This issue is unfortunately not the only one introduced by Pylint 2. Just from looking at the number of opened issues in the last couple of weeks, it seems like Pylint 2 fails to properly understand a lot of more complex code structures and as a result it reports a lot of false warnings and errors.

To me, the most surprising part is that such code structures were actually understood correctly by Pylint 1.

**Comment by PCManticore** [comment]
Hey folks, thanks for creating an issue!

So what happened here is that with https://github.com/PyCQA/pylint/commit/9a9de854741e2f3d8f97cbace3d7de444acae340, the `assignment-from-no-return` and `assignment-from-none` got the capability to look for bound methods as well, not just for functions as it was doing previously. But what you both hit was the fact that `pylint` doesn't currently understand decorators when inferring a value. This is an old issue that we want to address eventually but didn't get to it just yet. Since we already do this for a couple of other checks, I just pushed a fix to simply ignore decorated methods/functions, as there's no way knowing that they might return something without looking at the decorator itself (which will eventually happen though)

Also I'd like to say that with every release `pylint` and subsequently `astroid` become smarter, not dumber :) It's just that with every new inference capability that we add to `astroid` there might be a chance of that backfiring, since now we might need some other capability for understanding Python better (e.g. adding support for understanding decorators might still need support for understanding descriptors and so on)
